Add new address discovery schemes for testing and benchmarking

rvl opened the pull request:
Relates to issue #100

# Overview

This adds supplementary address discovery schemes for benchmarking.

Some things related to creating wallets needed to be moved around to accommodate.

parsonsmatt commented:
We're getting a test error because we're not passing the required parameters. Any idea on what stuff we should use for that?

KtorZ commented on src/Cardano/Wallet/API/Internal/Handlers.hs:
```diff
@@ -12,25 +17,53 @@ import           Cardano.Wallet.API.V1.Types (V1, Wallet, WalletImport)
+            case servantError of
+                FailureResponse (Response (Status statusCode _) _ _ _)
+                    | statusCode == 404 ->
+                        throwError err404
```
I wonder if the intent wouldn't be a bit clearer if we go from `ClientError () -> WalletError` instead, and go with `throwError $ fromNodeError err`. In the end, this function is doing nothing else than a mapping between node errors and our wallet errors.

parsonsmatt replied:
Well, the `ClientError` type is basically just `ServantError` from the Client with an extra slot for some other type of error. Do we have a mapping for that already?

KtorZ replied:
Not yet.

KtorZ commented on src/Cardano/Wallet/Server/CLI.hs:
```diff
@@ -49,21 +49,30 @@ data WalletDBOptions = WalletDBOptions {
     -- ^ Whether or not to run the monitoring API.
-    , monitoringApiPort   :: !Word16
+    , monitoringApiPort       :: !Word16
     -- ^ The port the monitoring API should listen to.
```
Gotta get rid of those bits about `monitoring` at some point since we just replaced it with the node API.

parsonsmatt replied:
Neat, we can mark them as deprecated on the CLI and deprecate the fields :)

KtorZ commented on src/Cardano/Wallet/Server/Plugins.hs:
```diff
@@ -79,13 +85,32 @@ apiServer
+            $ "There was an error connecting to the node: "
+            <> show err
+        Right _ -> do
+            logInfo "The nod responded successfully."
```
*node

KtorZ commented on stack.yaml:
```diff
@@ -1,10 +1,32 @@
+  # - ../cardano-sl/util
+  # - ../cardano-sl/util/test
+  # - ../cardano-sl/utxo
+  # - ../cardano-sl/x509
```
Hmmm... I do prefer the commit ref approach to be honest. Stack sometimes rebuild packages for no reasons whereas, updating the commit ref in the snapshot does trigger a new build only once. I'd rather not have that on the repo.

parsonsmatt replied:
Ok, I'll remove it. Commit ref is fine but it requires several more steps to try code.

KtorZ commented on src/Cardano/Wallet/Server/Plugins.hs:
```diff
@@ -24,15 +25,20 @@ import           Data.Aeson (encode)
 import qualified Data.ByteString.Char8 as BS8
 import qualified Data.Text as T
 import           Data.Typeable (typeOf)
+import           Data.X509.Validation (HostName)
```
NOTE: `HostName` is just a type-alias for string. Not sure it's worth pulling in a new dep for this :( ..

parsonsmatt replied:
Good catch! I'll revert that.

KtorZ commented on src/Cardano/Wallet/Server/CLI.hs:
```diff
@@ -157,6 +170,30 @@ walletBackendParamsParser = WalletBackendParams <$> enableMonitoringApiParser
+    tlsCaCertPathParser = strOption
+        $ long "tls-ca-cert"
+        <> metavar  "FILEPATH"
+        <> help "Path to TLS CA public certificate"
```
I think we could clarify the `help` messages for all those TLS parsers (the original `tlsParamParser` included) to remove confusion between which is for what purpose.

parsonsmatt replied:
Good point :)

KtorZ commented on src/Cardano/Wallet/Server/CLI.hs:
```diff
@@ -49,21 +49,30 @@ data WalletDBOptions = WalletDBOptions {
+    , walletRunMode           :: !RunMode
     -- ^ The mode this node is running in.
-    , walletDbOptions     :: !WalletDBOptions
+    , walletDbOptions         :: !WalletDBOptions
```
Gotta love those diffs :heart_eyes:  .... :no_mouth: ... Who was in favor of variable vertical alignment again :smile: ?

parsonsmatt replied:
Apparently if you had the `?w=1` thing then GitHUb renders it without whitespace changes. But ths isn't a UI option anywhere that I can find :\

KtorZ replied:
TIL. Good one.

parsonsmatt commented:
Got it running thanks to help from @KtorZ , now getting these test failures:

```
Failures:

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  1) Accounts only account's balance can be retrieved (non empty balance)
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Accounts/only account's balance can be retrieved (non empty balance)/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  2) Transactions successful payment appears in the history
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/successful payment appears in the history/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  3) Transactions estimate fees of well-formed transaction returns
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/estimate fees of well-formed transaction returns/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  4) Transactions payment fails when wallet has insufficient funds
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/payment fails when wallet has insufficient funds/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  5) Transactions payment fails when wallet cannot cover for fee
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/payment fails when wallet cannot cover for fee/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  6) Transactions balance is increased when valid key is redemeed
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/balance is increased when valid key is redemeed/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  7) Transactions Reg#141: metadata become inconsistent after a wallet is deleted
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/Reg#141: metadata become inconsistent after a wallet is deleted/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  8) Transactions Reg#141: metadata become inconsistent after an account is deleted
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/Reg#141: metadata become inconsistent after an account is deleted/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  9) Transactions Reg#141: metadata becomes inconsistent after a wallet is deleted and restored
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/Reg#141: metadata becomes inconsistent after a wallet is deleted and restored/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  10) Transactions coin selection: no extra inputs, no extra change
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/coin selection: no extra inputs, no extra change/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  11) Transactions coin selection: empties a wallet
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/coin selection: empties a wallet/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  12) Transactions coin selection: needs one extra input
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/coin selection: needs one extra input/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  13) Transactions coin selection: needs many extra inputs
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Transactions/coin selection: needs many extra inputs/"

  test/integration/Test/Integration/Framework/DSL.hs:142:23: 
  14) Wallets UTxO statistics reflect wallet's activity
       uncaught exception: IOException of type UserError
       user error (expectTxStatusEventually: waited too long for statuses.)

  To rerun use: --match "/Wallets/UTxO statistics reflect wallet's activity/"

Randomized with seed 804300954

Finished in 856.4199 seconds
25 examples, 14 failures
```

akegalj commented:
@parsonsmatt that output looks very similar (same?) what I have encountered 2 weeks ago https://input-output-rnd.slack.com/archives/GBT05825V/p1544697247470000?thread_ts=1544689822.455300&cid=GBT05825V

akegalj commented:
@parsonsmatt how did you fix it? 

> The tests are passing now if you want to re-review :)

parsonsmatt commented:
@akegalj The `init tls` line in the core node fixed it. How or why, I have no idea.

KtorZ commented on src/Cardano/Wallet.hs:
```diff
@@ -118,36 +119,46 @@ data NewWallet = NewWallet
-    } deriving (Show, Generic)
+    , addressDiscoveryConfig
+        :: AddressDiscoveryConfig s
+    }
```
A suggestion here: let's remove the `NewWallet` type wrapper here and simply modify the wallet layer signature as follows:

```hs
        :: WalletName
        -> Key scheme 'Root0 XPrv
        -> Passphrase "encryption"
        -> s
        -> ExceptT (ErrWalletAlreadyExists "createWallet") IO WalletId
```

Leave the creation of the state to the caller :+1:

rvl replied:
OK good idea. Turns out `NewWallet` is kind of helpful, so I have added a `createWalletSeq :: WalletLayer SeqState -> NewWallet ExceptT ErrWalletAlreadyExists IO WalletId`.

KtorZ commented on src/Cardano/Wallet.hs:
```diff
@@ -118,36 +119,46 @@ data NewWallet = NewWallet
+    initState :: NewWallet s -> (WalletId, s)
+
+instance InitState SeqState where
+    initState w = (wid, seqState)
```
We won't require a type-class here (such abstraction for the wallet scheme is actually hard to have in practice because the random cardano wallet and the sequential have rather different and not-so-compatible setup :(), but rather, let's make use of a helper function which yield a `SeqState` from a seed, generation passphrase and address pool gap.

```hs
initSeqState 
    :: (Passphrase "seed", Passphrase "generation") 
    -> AddressPoolGap 
    -> (Key 'Seq Root0 XPrv, SeqState)
```

This will play nicely with the previous comment and should be also flexible enough for testing :+1:

rvl replied:
OK added, but the testing AD scheme(s) don't have any key. So I let the `initSeqState` function calculate the wallet ID.

KtorZ committed:
make the wallet layer state-agnostic (instead of requiring a 'SeqState')

KtorZ committed:
Add stress restoration benchmarks

This creates a new 'special' address scheme that accept a given
proportion of addresses on the chain (e.g. 50%). This allows us to
artificially create very big wallets and observe how the wallet backend
survive and behave.

Co-authored-by: Rodney Lorrimar <rodney.lorrimar@iohk.io>

KtorZ committed:
add extra property tests for the 'putPrivateKey' DB function

KtorZ committed:
slightly review README file (move testing bits to Wiki & document benchmark)

KtorZ reviewed (approved):
reviewed & rebased on top of `master`
